### DISCUSSION TITLE
Making ZSH work nicely with `[` and `[[`

### DISCUSSION THREAD

**Discussion by targodan** [binref/refinery#18] [Show and tell]
Frames don't work well with zsh out-of-the-box, since zsh interprets the double square brackets `[[`. It errors with `zsh: bad pattern: [[`.

One *could* just escape the brackets, but I found a workaround for this. My workaround consists of a little python script, with code I stole from your setup.py and a small shell script for zsh.

## list-entrypoints.py

```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
from typing import Dict, List, Optional, Set, Tuple

import os
import refinery

__prefix__ = os.getenv('REFINERY_PREFIX') or ''


def get_entrypoints() -> List[Tuple[str, str]]:
    refinery._cache.reload()

    def normalize_name(name: str, separator: str = '-'):
        return separator.join([segment for segment in name.split('_')])

    if __prefix__ == '!':
        entrypoints = []
    else:
        entrypoints = [
            (F'{__prefix__}{normalize_name(name)}', F'{path}:{name}.run')
            for name, path in refinery._cache.units.items()
        ]
    entrypoints.append(("binref", "refinery.explore:explorer"))
    return entrypoints


for name, _ in get_entrypoints():
    print(name)
```

## init.sh

```zsh
#!/usr/bin/zsh

for entrypoint in $(python list-entrypoints.py); do
    alias $entrypoint="noglob $entrypoint"
done
```

## Usage

When I want to use binary-refinery, I now just do `source init.sh` and the noglob-alias handles this issue.

Feel free to use these snippets. If you want, you can also include these scripts in the repo and/or docs. :blush:

**Comment by huettenhain** [comment]
Awesome! I think there is a slightly more concise way to implement `list-entrypoints.py`, without assuming that the `REFINERY_PREFIX` environment variable is still defined at the time the script is executed:
```python
import pkg_resources
for ep in pkg_resources.iter_entry_points('console_scripts'):
    if ep.module_name.startswith('refinery'):
        print(ep.name)
```
Could you check if this works well for you? I am contemplating how to best add this to the documentation. Options I see:

1. Adding this to the README in-line, i.e. add a small section that shows the two small scripts and explains how to use them.
2. Adding a top-level subdirectory where the scripts are placed with their own README, then add a section in the top-level README that references it. With this option, we could even add a small table of known shells and how well they are supported. 

I am currently gravitating towards the second option. What do you think?

**Comment by targodan** [comment]
Nice, your code is indeed much nicer. :blush: (I'm not that deep into python packaging magic).

I went on to glue this into a single shell script. I think that would make it easier to copy-paste just a single script, rather than two files.

```zsh
#!/usr/bin/zsh

function alias-noglob {
    while read -r entrypoint; do
        alias $entrypoint="noglob $entrypoint"
    done
}

python <<EOF | alias-noglob
import pkg_resources
for ep in pkg_resources.iter_entry_points('console_scripts'):
    if ep.module_name.startswith('refinery'):
        print(ep.name)
EOF
```

About where to put it: I also think the second option is best. That would even allow you to put the shell script itself into that subdirectory, allowing folks to grab it via raw.github.com. Adding yet another alternative to copy-pasting.

**Comment by huettenhain** [comment]
Sorry for the delay! If you want at least some of your duly deserved credit, you could open a PR that adds the file to a new root directory called "zsh". I'll add some README text around it after. If you prefer that I add it myself, just let me know.

**Reply by targodan** [reply]
No problem at all. 😊

I'd be happy to open a pull request. Though I might not get around to it till Monday.

I also have been fiddling around with writing some VirusTotal and jq units (still work in progress) so you'll probably hear from me again. 😉

**Reply by targodan** [reply]
Sorry for the delay, the PR is now open. :)

**Reply by jhhcs** [reply]
merged!

**Comment by huettenhain** [comment]
Awesome! Although I'll warn you that I will probably not accept anything as a unit that does network callouts, but feel free to try and change my mind ;-).
